linker-interface/shared/src/main/scala/org/scalajs/linker/interface/OutputSpec.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.interface
+
+final class OutputSpec private (
+    val directory: OutputDirectory,
+    val jsName: ModuleSpec.ModuleID => String,
+    val sourceMapName: ModuleSpec.ModuleID => String,
```

Comment:
I'm unsure where to put `jsName` / `sourceMapName`. If I put it here, it is nice for the sbt plugin, since it keeps full control. OTOH, we need more machinery in the sbt plugin to support the "mjs" suffix for node.

Maybe putting these into the config (and then maybe just make them suffix strings due to serializability) and instead return a `LinkingReport` containing the files that were written might be a better idea.

Reply:
I'm less and less convinced this is the right spot TBH. I think we should have patterns for the following:

- JS file name
- Source map name
- Source map URI
- JS file URI
- module name (for include)

At that point, it really doesn't feel like this belongs to `OutputSpec`, but rather in a `ModulePatterns` option class in `StandardConfig`. WDYT?

Reply:
Hmmm... Turns out like this we cannot easily provide a backwards compatible method in `Linker` (because `config` cannot be change at that point).

I'm unsure how to proceed. IMHO as soon as things modify the *contents* of the files, it definitely should go into `config` :-/

Reply:
Perhaps one way to answer those questions: `StandardConfig` is just *one* way of configuring a linker, i.e., a standard linker. Does it make sense for another linker to be configured in a different way for file names? Would the sbt plugin be able to survive a different `scalaJSLinker` implementation that used a completely different scheme for file name configuration?

Today, the sbt plugin considers `scalaJSLinkerConfig` *almost* entirely as a blackbox, only used to create the default `scalaJSLinkerConfig`. The "almost" is because we actually dig into its `moduleKind` field to build `jsEnv` (and having some weird warning that if `moduleKind` differs `in fast/fullOptJS` than just in the configuration, things will break). To me this suggests that `moduleKind` already has a very specific status, and maybe shouldn't have been part of the `scalaJSLinkerConfig` to begin with. If these module patterns would equally affect the rest of the sbt plugin, perhaps they should stay out of `StandardConfig`?

Reply:
Hah, I think this is an excellent way of looking at it.

So what we observe is that the sbt plugin is peeking at the `scalaJSLinkerConfig` to find out which `moduleKind` the output is. It does this without taking the `scalaJSStage` into account, requiring the weird warning.

IMHO a better way to deal with this is a `LinkingReport` providing this information. At this point, the sbt plugin does not need to know/care anymore *why* a certain module type was returned (we could imagine a custom linker that only supports a specific type). It simply knows *which* module type was returned.

A similar strategy can then be applied to output files: The `LinkingReport` will simply list all the file names (and source maps) that have been produced.

This partially also addresses the backwards compatibility issue: We need not care about the config for file names but can
- write to a in memory directory
- retrieve the single output from the `LinkingReport` and write to the actual output

It does not address how to deal with `sourceMapURI` / `jsFileURI`.

Reply:
How about something like this?

```scala
final class OutputPatterns private (
    val jsFile: String,
    val sourceMapFile: String,
    val moduleName: String,
    val jsFileURI: String,
    val sourceMapURI: String
) {
  def this() = {
    this(
        jsFile = "[module-id].js",
        smFile = "[js-file].map",
        moduleName = "./[js-file]",
        jsFileURI = "[js-file]",
        smFileURI = "[sm-file]"
    )
  }

  // snip
}
```

This is inspired by webpack's placeholders. We could make this fully backwards compatible except:
- There would always be a link between source-map / js file.
- `jsFileURI` and `sourceMapURI` of a given `LinkerOutput` are set to the same (non-empty) value.

We'd have to combine this with a `LinkingReport` such that the sbt-plugin can recover root files and types.

@sjrd WDYT?